src/cargo/util/semver_ext.rs

```diff
@@ -111,15 +121,13 @@ impl OptVersionReq {
         }
     }
 
-    /// Since Semver does not support prerelease versions,
-    /// the simplest implementation is taken here without comparing the prerelease section.
-    /// The logic here is temporary, we'll have to consider more boundary conditions later,
-    /// and we're not sure if this part of the functionality should be implemented in semver or cargo.
+    /// Allow to match any "Semver-compatible" pre-release version.
+    /// The matches_prerelease semantics see [`semver_eval_ext`].
     pub fn matches_prerelease(&self, version: &Version) -> bool {
         if version.is_prerelease() {
-            let mut version = version.clone();
-            version.pre = semver::Prerelease::EMPTY;
-            return self.matches(&version);
+            if let OptVersionReq::Req(req) = self {
+                return req.matches_prerelease(version);
+            }
         }
         self.matches(version)
```

Comment:
Should we remove the `is_prerelease` check? I worry there might be some weird corner case we might mask and it seems like it shouldn't be relevant anymore

Reply:
OK